pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,184 @@
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	issuerType, err := apiutil.NameForIssuer(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		c.log.Error(err,
+			"issuer-name", issuerObj.GetName())
+		return nil
+	}
+
+	// This CertificateRequest is not meant for us, ignore
+	if issuerType != c.issuerType {
+		c.log.V(5).Info("issuer reference type does not match resource kind, ignoring",
+			"certificaterequest-issuer-type", issuerType,
+			"issuer-type", c.issuerType)
+		return nil
+	}
+
+	el := validation.ValidateCertificateRequest(crCopy)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	i, err := c.issuerFactory.IssuerFor(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerInit, "Internal error initialising issuer: %v", err)
+		return nil
+	}
+
+	dbg.Info("Fetched issuer resource referenced by certificate request", "issuer_name", crCopy.Spec.IssuerRef.Name)
+
+	if len(cr.Status.Certificate) == 0 {
+		dbg.Info("Invoking sign function as existing certificate does not exist")
+		return c.sign(ctx, crCopy, i)
+	}
+
+	dbg.Info("Update certificate request status if required")
+	c.setCertificateRequestStatus(crCopy)
+
+	// TODO: Metrics??
+
+	dbg.Info("Certificate does not need updating.")
+
+	return nil
+}
+
+// return an error on failure. If retrieval is succesful, the certificate data
+// will be stored in the certificate request status
+func (c *Controller) sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuer issuer.Interface) error {
+	log := logf.FromContext(ctx)
+
+	defer c.setCertificateRequestStatus(cr)
+
+	resp, err := issuer.Sign(ctx, cr)
+	if err != nil {
+		log.Error(err, "error issuing certificate request")
+		return err
+	}
+
+	// if the issuer has not returned any data, exit early
+	if resp == nil {
+		return nil
+	}
+
+	if len(resp.Certificate) > 0 {
+		cr.Status.Certificate = resp.Certificate
+		cr.Status.CA = resp.CA
+
+		c.recorder.Event(cr, corev1.EventTypeNormal, successCertificateIssued, "Certificate issued successfully")
+	}
+
+	return nil
+}
+
+// setCertificateRequestStatus will update the status subresource of the
+// certificate request.
+func (c *Controller) setCertificateRequestStatus(cr *v1alpha1.CertificateRequest) {
+	// No cert exists yet
+	if len(cr.Status.Certificate) == 0 {
+		apiutil.SetCertificateRequestCondition(cr, v1alpha1.CertificateRequestConditionReady, v1alpha1.ConditionFalse,
+			errorCertificateNotExists, "Certificate does not exist")
+		return
+	}
+
+	// invalid cert
+	_, err := pki.DecodeX509CertificateBytes(cr.Status.Certificate)
+	if err != nil {
+		apiutil.SetCertificateRequestCondition(cr, v1alpha1.CertificateRequestConditionReady,
+			v1alpha1.ConditionFalse, errorCertificateParse, "Failed to decode certificate PEM")
```

Comment:
We need to be careful here, especially around backing off and timing out. If the upstream CA continuously returns invalid data, we will continuously retry due to this check, potentially hitting rate limits (or credit card limits!!).

I'd lean towards saying that we should mark this certificate request as 'failed' in a more permanent way here, and allow higher-level consumers of this API (e.g. the Certificates controller) to handle retrying, as we have some existing mechanisms for that already. I'm not 100% sure though, so would like to hear your thoughts 😄

Reply:
Yeah I agree, that sounds right to me.

I guess this is already the current behavior in that sign is only called when `len(cr.Status.Cert) == 0` but I think changing the msg to a hard _fail_ is more clear that no more action will be taken by the controller.

Reply:
The issue I see is that the `cr.Status.Cert` length will still be zero, as we don't actually set this field to the invalid data from what I can see? It may be best to persist the invalid data, but mark the certificate as "NotReady". This avoids having to maintain a separate `state` field.